sql: support JSONB subscripting for SELECT cases

otan opened the pull request:
I've kept the `ArraySubscript` class names and references as PG does
this too. We can choose to change this in later iterations.

Refs [#77434](https://github.com/cockroachdb/cockroach/issues/77434)

Release note (sql change): Added support for JSONB subscripting in
SELECT-style cases, e.g. SELECT json_field['a'] ... WHERE
json_field['b'] = ...

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/82877)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
just had small comments; maybe someone from sql-queries should look too

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-:-N4h4ibgEz1h75U6eY-z:b-yyp5wz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[`pkg/sql/opt/memo/typing.go` line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N4cf7C61p3vJcHBdVfw:-N4cf7C61p3vJcHBdVfx:bbc7hmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7f83979061fd4c43b39f3c1cc044789afd67bc8/pkg/sql/opt/memo/typing.go#L255)):*
> ```Go
> 	switch t.Family() {
> 	case types.JsonFamily:
> 		return t
> ```

i don't follow this part. do you think we should update the comment on the function?
___
*[`pkg/sql/opt/norm/fold_constants_funcs.go` line 506 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N4cfwQG4-0n2AmJ4bnj:-N4cfwQG4-0n2AmJ4bnk:b-2zgmd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7f83979061fd4c43b39f3c1cc044789afd67bc8/pkg/sql/opt/norm/fold_constants_funcs.go#L506)):*
> ```Go
> 		if resolvedType.Family() == types.ArrayFamily {
> 			resolvedType = resolvedType.ArrayContents()
> 		}
> ```

nit: should it return an error if it's not an array or json?
___
*[`pkg/sql/opt/optbuilder/scalar.go` line 176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N4h3t6s3_WQvMTcFlAr:-N4h3t6x7YjYBzsHWb-W:bd6y0tu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5f7b03a7b3b15fe5a4894d8b0f0282a488cbd1/pkg/sql/opt/optbuilder/scalar.go#L176)):*
> ```Go
> 
> 		if len(t.Indirection) != 1 {
> 			panic(unimplementedWithIssueDetailf(32552, "ind", "multidimensional indexing is not supported"))
> ```

would we still want this error message to be shown for arrays? otherwise, it'd hit the assertion error in EvalIndirection, right?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
maybe @mgartner ?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-:-N4hNiLfEnlND5zB--8q:bfyi7g4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/opt/memo/typing.go` line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N4cf7C61p3vJcHBdVfw:-N4hMmAeBMvvrD2LEX-x:b-pxaxmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7f83979061fd4c43b39f3c1cc044789afd67bc8/pkg/sql/opt/memo/typing.go#L255)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i don't follow this part. do you think we should update the comment on the function?
</blockquote></details>

comment updated and hopefully it makes more sense.
___
*[`pkg/sql/opt/optbuilder/scalar.go` line 176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N4h3t6s3_WQvMTcFlAr:-N4hNeeH3a7ySK0G7oE6:by2glrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5f7b03a7b3b15fe5a4894d8b0f0282a488cbd1/pkg/sql/opt/optbuilder/scalar.go#L176)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

would we still want this error message to be shown for arrays? otherwise, it'd hit the assertion error in EvalIndirection, right?
</blockquote></details>

i think it gets caught elsewhere, but added it just in case.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82877)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/opt/optbuilder/scalar.go` line 176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N4h3t6s3_WQvMTcFlAr:-N4hSgkDBNT2ZNqi-9GZ:b-y1hrvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5f7b03a7b3b15fe5a4894d8b0f0282a488cbd1/pkg/sql/opt/optbuilder/scalar.go#L176)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i think it gets caught elsewhere, but added it just in case.
</blockquote></details>

actually, there is no easy way to check for this. it does have checks further down the stack though:

```
demo@127.0.0.1:26257/defaultdb> select array['1', '2'][1][2];
ERROR: unimplemented: multidimensional indexing: ARRAY['1', '2'][1][2]
SQLSTATE: 0A000
HINT: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/32552/v22.2

demo@127.0.0.1:26257/defaultdb> create table t (a string[]);
CREATE TABLE


Time: 18ms total (execution 18ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> insert into t values (array['1', '2']);
INSERT 0 1


Time: 12ms total (execution 12ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> select t[1][2] from t;
ERROR: cannot subscript type tuple{string[] AS a} because it is not an array or json object
SQLSTATE: 42804
demo@127.0.0.1:26257/defaultdb> select a[1][2] from t;
ERROR: unimplemented: multidimensional indexing: a[1][2]
SQLSTATE: 0A000
HINT: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/32552/v22.2
````


<!-- Sent from Reviewable.io -->

otan commented:
*[`pkg/sql/opt/optbuilder/scalar.go` line 176 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N4h3t6s3_WQvMTcFlAr:-N4hrVDX559MjokrXnBW:bcpvush) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a5f7b03a7b3b15fe5a4894d8b0f0282a488cbd1/pkg/sql/opt/optbuilder/scalar.go#L176)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

actually, there is no easy way to check for this. it does have checks further down the stack though:

```
demo@127.0.0.1:26257/defaultdb> select array['1', '2'][1][2];
ERROR: unimplemented: multidimensional indexing: ARRAY['1', '2'][1][2]
SQLSTATE: 0A000
HINT: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/32552/v22.2

demo@127.0.0.1:26257/defaultdb> create table t (a string[]);
CREATE TABLE


Time: 18ms total (execution 18ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> insert into t values (array['1', '2']);
INSERT 0 1


Time: 12ms total (execution 12ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> select t[1][2] from t;
ERROR: cannot subscript type tuple{string[] AS a} because it is not an array or json object
SQLSTATE: 42804
demo@127.0.0.1:26257/defaultdb> select a[1][2] from t;
ERROR: unimplemented: multidimensional indexing: a[1][2]
SQLSTATE: 0A000
HINT: You have attempted to use a feature that is not yet implemented.
See: https://go.crdb.dev/issue-v/32552/v22.2
````
</blockquote></details>

uhh, nvm


<!-- Sent from Reviewable.io -->

mgartner reviewed (changes requested):
Very cool! We should probably normalize these to `->` expressions when possible so that they can be index accelerated. I'll create an issue to track.

> Reviewed 7 of 10 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-:-N5HbkVV2ncrJDVio7tu:b-b7r32m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[`pkg/sql/opt/norm/fold_constants_funcs.go` line 506 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N5Ha8Dm04y9CsgNxCkv:-N5Ha8Dm04y9CsgNxCkw:blc9xi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1db5511909b6f3b7622d4a11b51882512b5abdb6/pkg/sql/opt/norm/fold_constants_funcs.go#L506)):*
> ```Go
> 		switch input.DataType().Family() {
> 		case types.JsonFamily:
> 			resolvedType = input.DataType()
> ```

You should add some tests for folding JSON indirections in this section: https://github.com/cockroachdb/cockroach/blob/a134352e19e4275b019081bd1944587e1b29abf0/pkg/sql/opt/norm/testdata/rules/fold_constants#L1122-L1216
___
*[`pkg/sql/sem/tree/type_check.go` line 636 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N5Hb1Aw88a2ywBzk_GM:-N5Hb1Aw88a2ywBzk_GN:bnfvnpi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1db5511909b6f3b7622d4a11b51882512b5abdb6/pkg/sql/sem/tree/type_check.go#L636)):*
> ```Go
> 		for _, t := range expr.Indirection {
> 			if t.Slice {
> 				return nil, pgerror.Newf(pgcode.DatatypeMismatch, "cannot reference a slice with JSON")
> ```

nit: is "reference" the right word in this error message? Is "index" is more appropriate?
___
*[`pkg/sql/logictest/testdata/logic_test/json` line 911 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N5HbKjS-ZT-Tl3-suFz:-N5HbKjTDVrpaJeJMlTE:b38qdoe) ([raw file](https://github.com/cockroachdb/cockroach/blob/1db5511909b6f3b7622d4a11b51882512b5abdb6/pkg/sql/logictest/testdata/logic_test/json#L911)):*
> ```
> NULL  NULL  NULL  NULL
> 
> # Error cases.
> ```

nit: add a test case with slice indexing


<!-- Sent from Reviewable.io -->

mgartner commented:
> ... We should probably normalize these to -> expressions when possible so that they can be index accelerated. I'll create an issue to track.

Done: https://github.com/cockroachdb/cockroach/issues/83441

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82877)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @otan, and @rafiss)

___
*[`pkg/sql/sem/tree/type_check.go` line 636 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N5Hb1Aw88a2ywBzk_GM:-N5atEHs-NotrJvHFJiV:b-sgr0lz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1db5511909b6f3b7622d4a11b51882512b5abdb6/pkg/sql/sem/tree/type_check.go#L636)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: is "reference" the right word in this error message? Is "index" is more appropriate?
</blockquote></details>

replaced with `jsonb subscript does not support slices` a la PG


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82877)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rafiss)

___
*[`pkg/sql/opt/norm/fold_constants_funcs.go` line 506 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N5Ha8Dm04y9CsgNxCkv:-N5auqKq9qhpreh2qQof:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1db5511909b6f3b7622d4a11b51882512b5abdb6/pkg/sql/opt/norm/fold_constants_funcs.go#L506)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

You should add some tests for folding JSON indirections in this section: https://github.com/cockroachdb/cockroach/blob/a134352e19e4275b019081bd1944587e1b29abf0/pkg/sql/opt/norm/testdata/rules/fold_constants#L1122-L1216
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
sql: support JSONB subscripting for SELECT cases

I've kept the `ArraySubscript` class names and references as PG does
this too. We can choose to change this in later iterations.

Release note (sql change): Added support for JSONB subscripting in
SELECT-style cases, e.g. SELECT json_field['a'] ... WHERE
json_field['b'] = ...

otan commented:
*[`pkg/sql/opt/norm/testdata/rules/cycle` line 6 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N5av5HC3Gp5QuRAfQAL:-N5av5HC3Gp5QuRAfQAM:bsb4k7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a9ec7c691ab2fbfc33a79867844ee3a4a4165b2/pkg/sql/opt/norm/testdata/rules/cycle#L6)):*
> ```
> (NormCycleTestRel (True))
> ----
> error: optimizer factory constructor call stack exceeded max depth of 10000
> ```

(this `-rewrite`s for me on my M1 macbook, reverted)


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/norm/testdata/rules/cycle` line 6 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N5av5HC3Gp5QuRAfQAL:-N5b0sIJ42TPgyo0qjAF:baecslr) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a9ec7c691ab2fbfc33a79867844ee3a4a4165b2/pkg/sql/opt/norm/testdata/rules/cycle#L6)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

(this `-rewrite`s for me on my M1 macbook, reverted)
</blockquote></details>

That's odd and unsettling. Is it somehow optimizing the infinite recursion away? Is that possible?


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/opt/norm/testdata/rules/cycle` line 6 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N5av5HC3Gp5QuRAfQAL:-N5eq0HC6ieJJiqn9vuZ:b-jyhess) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a9ec7c691ab2fbfc33a79867844ee3a4a4165b2/pkg/sql/opt/norm/testdata/rules/cycle#L6)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

That's odd and unsettling. Is it somehow optimizing the infinite recursion away? Is that possible?
</blockquote></details>

I created an issue to track: https://github.com/cockroachdb/cockroach/issues/83512


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 4 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-:-N5evn7M3Hiaglz-nkAX:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[`pkg/sql/opt/norm/fold_constants_funcs.go` line 506 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/82877#-N5Ha8Dm04y9CsgNxCkv:-N5evkB-7wGaEUSJVAp4:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1db5511909b6f3b7622d4a11b51882512b5abdb6/pkg/sql/opt/norm/fold_constants_funcs.go#L506)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks!


<!-- Sent from Reviewable.io -->

otan commented:
bors r=mgartner
